File: app/src/main/java/com/karlnosworthy/examples/daggerv2/ApplicationImpl.java
package com.karlnosworthy.examples.daggerv2;

import android.app.Application;
import android.content.Context;


public class ApplicationImpl extends Application {

    private ApplicationComponent applicationComponent;


    public static ApplicationComponent using(Context context) {
        return ((ApplicationImpl) context.getApplicationContext()).getApplicationComponent();
    }

    @Override
    public void onCreate() {
        super.onCreate();
        applicationComponent = DaggerApplicationComponent.builder()
                .build();
    }

    private ApplicationComponent getApplicationComponent() {
        return applicationComponent;
    }
}

File: app/src/main/java/com/karlnosworthy/examples/daggerv2/ApplicationImpl.kt
package com.karlnosworthy.examples.daggerv2

import android.app.Application
import android.content.Context


class ApplicationImpl : Application() {

    private lateinit var applicationComponent: ApplicationComponent

    override fun onCreate() {
        super.onCreate()
        applicationComponent = DaggerApplicationComponent.builder()
                .build()
    }

    companion object {
        fun using(context: Context): ApplicationComponent {
            return (context.applicationContext as ApplicationImpl).applicationComponent
        }
    }
}
